### ISSUE TITLE
OpenSSL 1.1.0 on armv7 device

### ISSUE THREAD

**Issue by dingosky** [x2on/OpenSSL-for-iPhone#131]
Apps using builds for OpenSSL 1.1.0 fail to run on armv7 and armv7s devices.

I had this issue in my own app, but to reproduce I used this repo's example project. I verified issue with all four current releases of OpenSSL 1.1.0. Each runs fine in simulator and on arm64 device, but fail during launch on armv7 device.

Here's the log from one such failure using the example Xcode project:

```
Incident Identifier: 41A7B41A-D822-460F-ACEE-41A0A5509193
CrashReporter Key:   a8bdc331004cff1b21d23af423c224be559fb4ef
Hardware Model:      iPhone5,1
Process:             OpenSSL [468]
Path:                /private/var/containers/Bundle/Application/3D26DBBE-67CC-4A86-B250-6EAC9E591BBF/OpenSSL.app/OpenSSL
Identifier:          de.felixschulze.OpenSSL
Version:             1 (3.1)
Code Type:           ARM (Native)
Role:                Unspecified
Parent Process:      launchd [1]
Coalition:           de.felixschulze.OpenSSL [434]


Date/Time:           2016-11-24 10:56:49.2936 -0800
Launch Time:         2016-11-24 10:56:49.0000 -0800
OS Version:          iPhone OS 10.1.1 (14B150)
Report Version:      104

Exception Type:  EXC_BAD_ACCESS (SIGBUS)
Exception Subtype: EXC_ARM_DA_ALIGN at 0x00000101
Termination Signal: Bus error: 10
Termination Reason: Namespace SIGNAL, Code 0xa
Terminating Process: exc handler [0]
Triggered by Thread:  0

Filtered syslog:
None found

Thread 0 name:  Dispatch queue: com.apple.main-thread
Thread 0 Crashed:
0   libswiftObjectiveC.dylib      	0x002734f4 0x26c000 + 29940
1   dyld                          	0x002addde ImageLoaderMachO::doModInitFunctions(ImageLoader::LinkContext const&) + 300
2   dyld                          	0x002adf50 ImageLoaderMachO::doInitialization(ImageLoader::LinkContext const&) + 20
3   dyld                          	0x002aa4e6 ImageLoader::recursiveInitialization(ImageLoader::LinkContext const&, unsigned int, char const*, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 366
4   dyld                          	0x002a9608 ImageLoader::processInitializers(ImageLoader::LinkContext const&, unsigned int, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 116
5   dyld                          	0x002a968c ImageLoader::runInitializers(ImageLoader::LinkContext const&, ImageLoader::InitializerTimingList&) + 52
6   dyld                          	0x0029cfb2 dyld::initializeMainExecutable() + 172
7   dyld                          	0x002a1388 dyld::_main(macho_header const*, unsigned long, int, char const**, char const**, char const**, unsigned long*) + 4004
8   dyld                          	0x0029c1d0 dyldbootstrap::start(macho_header const*, int, char const**, long, macho_header const*, unsigned long*) + 324
9   dyld                          	0x0029c040 _dyld_start + 64

Thread 0 crashed with ARM Thread State (32-bit):
    r0: 0x00000101    r1: 0x00000000      r2: 0x0016491c      r3: 0xfffff967
    r4: 0x00020254    r5: 0x00164878      r6: 0x0005234b      r7: 0x00404a8c
    r8: 0x3820f000    r9: 0x391548c4     r10: 0x002c951c     r11: 0x00000000
    ip: 0x1beec550    sp: 0x00404a60      lr: 0x000525a7      pc: 0x002734f4
  cpsr: 0x00000030

Binary Images:
0x45000 - 0x150fff OpenSSL armv7  <37103e09fd663ab3be9420c5cd561541> /var/containers/Bundle/Application/3D26DBBE-67CC-4A86-B250-6EAC9E591BBF/OpenSSL.app/OpenSSL
0x1c1000 - 0x1d0fff libswiftCoreGraphics.dylib armv7s  <c97df444e9c33889bc51c6eab0b442ca> /var/containers/Bundle/Application/3D26DBBE-67CC-4A86-B250-6EAC9E591BBF/OpenSSL.app/Frameworks/libswiftCoreGraphics.dylib
```
...elided...

**Comment by oneton** [comment]
I'm not sure what's going on, but this could be related to PR #126 
Unfortunately I do not have a device available to check this testcase. I tried to make my iPhone 4 work, but that leads to a whole different world of problems. :(
When the framework is built for multiple targets, there are some differences in configurations that end up in the opensslconf.h header file. The file used in the framework is the opensslconf.h from the first built target (by default iPhone simulator x64). If you compile the framework only for armv7s (use the --targets option, check --help for how to use it), do you still get the error?

**Comment by dingosky** [comment]
I noted the help listing regarding the include files being dependent on the order of the targets and had tried various adjustments to that order. I had not, however, built only for __armv7__ and did try that as you suggest, but to a similar result. Still crashing during system app initialization.

Here's a brief overview of what I've found:

 * The issue only occurs on __armv7__ and __armv7s__ devices. I've used
   * iPad mini, iOS 9.5.1
   * iPhone 7, iOS 10.1.1
 * The issue occurs whether __arm7v__ is the only target or in any position in the target list.
 * The issue occurs for all of the OpenSSL 1.1.0 releases, but not OpenSSL 1.0.2j
 * The crash occurs during image loader linking, before __AppDelegate.init()__
   is called. Execution never reaches any application specific code.

**Comment by anujseth** [comment]
Can confirm crash on an iPhone 5. Tried with the test app (trace below) and my own apps. 1.0.2j works fine, 1.1.0c crashes on launch. 1.1.0c does work fine on an iPad mini and all simulators. 

```
(lldb) bt
* thread #1: tid = 0x17dcf2, 0x002603e2, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=2, address=0x2603e2)
  * frame #0: 0x002603e2
    frame #1: 0x0003f5a6 OpenSSL`_armv7_neon_probe + 6
    frame #2: 0x1fe95664 dyld`ImageLoaderMachO::doModInitFunctions(ImageLoader::LinkContext const&) + 184
    frame #3: 0x1fe95764 dyld`ImageLoaderMachO::doInitialization(ImageLoader::LinkContext const&) + 20
    frame #4: 0x1fe92702 dyld`ImageLoader::recursiveInitialization(ImageLoader::LinkContext const&, unsigned int, char const*, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 382
    frame #5: 0x1fe91832 dyld`ImageLoader::processInitializers(ImageLoader::LinkContext const&, unsigned int, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 110
    frame #6: 0x1fe918b8 dyld`ImageLoader::runInitializers(ImageLoader::LinkContext const&, ImageLoader::InitializerTimingList&) + 52
    frame #7: 0x1fe85cd8 dyld`dyld::initializeMainExecutable() + 168
    frame #8: 0x1fe89c8a dyld`dyld::_main(macho_header const*, unsigned long, int, char const**, char const**, char const**, unsigned long*) + 3774
    frame #9: 0x1fe85226 dyld`dyldbootstrap::start(macho_header const*, int, char const**, long, macho_header const*, unsigned long*) + 398
    frame #10: 0x1fe85040 dyld`_dyld_start + 64
```

**Comment by dingosky** [comment]
@anujseth Yes, crash verified with **armv7** iPhone 5. Note the iPad mini crash I reported was **armv7** as well. I suspect the iPad mini that works for you is **arm64**.

Your trace is consistent with the final bullet in my previous post.

**Comment by anujseth** [comment]
@dingosky Yes, the iPad mini is `ARM-64 (Native)`

**Comment by swapnilgavali** [comment]
Yes, crash verified with armv7 iPhone 5. can u help me out for this issue

**Comment by carloscabanero** [comment]
Not sure there is an easy fix for this. This is the issue https://brewx.qualcomm.com/bws/content/gi/common/appseng/en/knowledgebase/docs/kb95.html, and I have previously fixed it on the ocb algorithm with https://github.com/blinksh/mosh/commit/dd1d82ed18a90e8c69f984498bf1ce53a618102c

Apparently the memcpy is even more efficient, on arm devices, but changing all instances doesn't look like an easy monkey patch :(

Maybe someone has a better idea :)

**Comment by oneton** [comment]
(Again, I don't have hardware to test, so I can only try to aid in the diagnosis)

The easiest way to proceed is to see whether the default OpenSSL build configuration for armv7 works. This should be a configuration that is supported by the OpenSSL team and can therefore be reported as an issue in case it is broken. This requires a few manual steps, but is still fairly straightforward. Please note that this results in an armv7 only build. (It should also work on armv7s)

- Download OpenSSL 1.1.0c or take the downloaded file from the script directory.
- Extract the file to a location of your choice (e.g. the src directory)
- Go to the main directory of the extracted archive (it contains the Configure script)
- Set a few environment variables required for the build:
`export CROSS_COMPILE=$(xcode-select -print-path)"/Toolchains/XcodeDefault.xctoolchain/usr/bin/"`
`export CROSS_TOP=$(xcode-select -print-path)"/Platforms/iPhoneOS.platform/Developer"`
`export CROSS_SDK="iPhoneOS"$(xcrun -sdk iphoneos --show-sdk-version)".sdk"`
- Run configure with a minimal set of options:
`./Configure ios-cross no-shared no-deprecated`
- Run make
`make -j8`
- Copy the built libraries (libcrypto.a and libssl.a) to the script's lib directory
- Open the example project in Xcode
- Check whether this works

If it doesn't, I reckon the problem is in the OpenSSL implementation and the best way forward would be to create a ticket for the OpenSSL repository on Github.

**Comment by dingosky** [comment]
I did the steps as outlined by @oneton (with an addition of embedding bitcode) and the crash on **armv7** still occurred. However, the suggestion led me to try a build script I previously created (with guidance from the OpenSSL team build script) that attempted to pare down **libcrypto** and skip **libssl** altogether via various `no-feature` options (feature being **des3**, etc) that didn't seem to work as I hoped. Anyway, using that build script OpenSSL 1.1.0c worked for all platforms, including **armv7** and **armv7s**. So I ferreted around to see what was different and saw that my build script included a **no-asm** configuration option. Based on that, I added an if statement to the `run_configure` function of this project's build script that appended **no-asm** to the **CONFIG_OPTIONS** env var if the architecture is **armv7** or **armv7s**. With that additional config option 1.1.0c runs on all platforms (I'm not doing tvOS though).

You could, of course, on a per build basis add **no-asm** to **CONFIG_OPTIONS**. I chose not to do that as it affects all architectures, not just the ones that seem to be problematic with regard to this issue. As such, I have not created a pull request of the above solution since it "breaks" the scripts current **Build options** output, which currently applies to all configurations and hence masks that the configuration for **armv7** and **armv7s** are different.

Finally, I do not make any claim to understanding any performance implications of disabling assembly code for the **armv7** and **armv7s** architectures.

TL;DR

Either pass in a **CONFIG_OPTION** of **no-asm** (applies to all architectures) or add an if directive to the build script for **no-asm** that applies only to **armv7** and **armv7s**.

**Comment by swapnilgavali** [comment]
thanks for all of us
i am compiling library using below link 
https://github.com/leenjewel/openssl_for_ios_and_android

**Comment by oneton** [comment]
Thank you for the detailed description! 

Please note the comment (quoted below) from one of the OpenSSL team members before considering the no-asm option as a permanent solution. 
> (...)
> And as **richsalz** implied, omitting assembly support is likely to be bad for security. It's even argued that disabling assembly support is plain incompatible with security by todays standards, especially on contemporary processors.
> (...)

See https://github.com/openssl/openssl/issues/2012 (second comment by **dot-asm**)

**Comment by dumganhar** [comment]
@oneton , I also get this crash error in armv7 | armv7s devices. What about the workaround if not use `no-asm` option?

**Comment by oneton** [comment]
I can only refer to my previous comment: try the diagnosis I described. If the build still fails to run with minimal configuration options, then describe the reproduction path and request the OpenSSL team for help at their Github repository.

**Comment by oneton** [comment]
So, status update.
After having another look at @anujseth's trace, I noticed that the referred `_armv7_neon_probe` exists in the OpenSSL source code. It is in the Perl assembly code in a section that was called on startup. Digging further, I found some things that could be related to the problem, so I reported my analysis to the OpenSSL team in openssl/openssl#2601.

The initial line of thought went in the direction of the bitcode argument (which is why I requested the reproduction without), but a second look did reveal a bug in the OpenSSL assembly code.

After very detailed and helpful comments with background information and an explanation of the problem, a fix was created today. I'd like to invite you to validate it. Since it was announced that OpenSSL 1.1.0e will be released on this Thursday, it might be included in that version if we can quickly confirm that the fix works.

To make it easy, I have created a version of the script that builds 1.1.0d with the supplied patch. It is available on [my Github repository](https://github.com/oneton/OpenSSL-for-iPhone/tree/test-armv7-patch) (use branch test-armv7-patch). Please note that it includes my suggested fix for #126 and that it shouldn't be used to build a different version.

**Comment by dingosky** [comment]
I built the test-armv7-patch branch and got the following error during debug installation on both an iPhone 5 running iOS 10.2 and an iPad mini (original) running iOS 9.3.5.

![screen shot 2017-02-13 at 4 21 13 pm](https://cloud.githubusercontent.com/assets/134293/22909665/40168e04-f20a-11e6-81bd-6c435882de4e.png)

The app on the iPhone 5 (i.e., it did actually install) worked fine afterward. The app on the iPad mini did not (nothing but a blank, black screen).

**Comment by oneton** [comment]
Please see the conversion in the issue on the OpenSSL repository (so I don't have to repeat everything I type ;)). What does the log file on the iPad mini say (Settings > Privacy > Diagnostics & Usage > Diagnostic & Usage Data > Entry starting with OpenSSL)?

**Comment by oneton** [comment]
Short update: an additional patch was added to skip the code that triggered the error in your screenshot. I pushed a new version to [the test branch on my fork](https://github.com/oneton/OpenSSL-for-iPhone/tree/test-armv7-patch). I'm particularly interested to know whether this new build runs on your iPad mini.

**Comment by dingosky** [comment]
I can confirm commit **bf386bd** on your forked test branch works on both my iPhone 5 and iPad mini (orig).

**Comment by oneton** [comment]
Thank you for testing! I've created a PR to update the script for OpenSSL 1.1.0e which includes the patch for this problem. Manually running the previous version with --version=1.1.0e will also do the trick. :)

**Comment by x2on** [comment]
So we could close this issue, or?

**Comment by oneton** [comment]
In short: yes.

**Comment by x2on** [comment]
Thanks a lot to all involved people for fixing this! 🥇
